`feature` and `command` decorators loose type annotations

If I use this patch on json-extension example:

```diff
diff --git a/examples/json-extension/server/server.py b/examples/json-extension/server/server.py
index efd6087..70fd3bf 100644
--- a/examples/json-extension/server/server.py
+++ b/examples/json-extension/server/server.py
@@ -90,7 +90,7 @@ def _validate_json(source):
 
 
 @json_server.feature(COMPLETION, trigger_characters=[','])
-def completions(params: CompletionParams = None):
+def completions(params: CompletionParams = None) -> CompletionList:
     """Returns completion items."""
     return CompletionList(False, [
         CompletionItem('"'),
diff --git a/examples/json-extension/server/tests/unit/test_features.py b/examples/json-extension/server/tests/unit/test_features.py
index b2ff762..3ec98d7 100644
--- a/examples/json-extension/server/tests/unit/test_features.py
+++ b/examples/json-extension/server/tests/unit/test_features.py
@@ -50,8 +50,9 @@ def _reset_mocks():
     server.show_message_log.reset_mock()
 
 
-def test_completions():
-    completion_list = completions()
+def test_completions() -> None:
+    completion_list = completions("type error")
+    reveal_type(completion_list)
     labels = [i.label for i in completion_list.items]
 
     assert '"' in labels
```

I would expect that mypy report type error for wrong `"type error"` argument and reveal  type of `completion_list` as `CompletionList`, instead, I have:

```
$ mypy examples/json-extension/server/tests/unit/test_features.py 
...
examples/json-extension/server/tests/unit/test_features.py:55: note: Revealed type is 'Any'
```

I would have expected something like:
```
$ mypy examples/json-extension/server/tests/unit/test_features.py
...
examples/json-extension/server/tests/unit/test_features.py:54: error: Argument 1 to "completions" has incompatible type "str"; expected "Optional[CompletionParams]"
examples/json-extension/server/tests/unit/test_features.py:55: note: Revealed type is 'pygls.types.CompletionList'
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning][semver].
 
 ## [Unreleased]
 
+### Changed
+
+- Fixed `@command`, `@feature` and `@thread` decorators to retain type of wrapped functions ([#89])
+
+[#89]: https://github.com/openlawlibrary/pygls/pull/89
+
 ### Added
 
 - Add comparisons and repr support to Range and Location types ([#90])

--- a/pygls/server.py
+++ b/pygls/server.py
@@ -21,7 +21,7 @@ import sys
 from concurrent.futures import Future, ThreadPoolExecutor
 from multiprocessing.pool import ThreadPool
 from threading import Event
-from typing import Callable, Dict, List
+from typing import Callable, Dict, List, TypeVar
 
 from pygls.types import (ApplyWorkspaceEditResponse, ConfigCallbackType, Diagnostic, MessageType,
                          RegistrationParams, TextDocumentSyncKind, UnregistrationParams,
@@ -33,6 +33,8 @@ from .types import ConfigurationParams
 from .workspace import Workspace
 
 logger = logging.getLogger(__name__)
+
+F = TypeVar('F', bound=Callable)
 
 
 async def aio_readline(loop, executor, stop_event, rfile, proxy):
@@ -234,7 +236,7 @@ class LanguageServer(Server):
         """Sends apply edit request to the client."""
         return self.lsp.apply_edit(edit, label)
 
-    def command(self, command_name: str) -> Callable:
+    def command(self, command_name: str) -> Callable[[F], F]:
         """Decorator used to register custom commands.
 
         Example:
@@ -244,7 +246,7 @@ class LanguageServer(Server):
         """
         return self.lsp.fm.command(command_name)
 
-    def feature(self, feature_name: str, **options: Dict) -> Callable:
+    def feature(self, feature_name: str, **options: Dict) -> Callable[[F], F]:
         """Decorator used to register LSP features.
 
         Example:
@@ -287,7 +289,7 @@ class LanguageServer(Server):
         """Sends message to the client's output channel."""
         self.lsp.show_message_log(message, msg_type)
 
-    def thread(self) -> Callable:
+    def thread(self) -> Callable[[F], F]:
         """Decorator that mark function to execute it in a thread."""
         return self.lsp.thread()
 

